docs/man/output.1
.\" Automatically generated by Pandoc 2.9.2.1
.\"
.TH "TRACKER-OUTPUT" "1" "2023/10" "" "Tracker Output Flag Manual"
.hy
.SS NAME
.PP
tracker \f[B]--output\f[R] - Control how and where output is printed
.SS SYNOPSIS
.PP
tracker \f[B]--output\f[R] <format[:file,\&...]> |
gotemplate=template[:file,\&...] | forward:url | webhook:url |
option:{stack-addresses,exec-env,relative-time,exec-hash,parse-arguments,parse-arguments-fds,sort-events}
\&...
.SS DESCRIPTION
.PP
The \f[B]--output\f[R] flag allows you to control how and where the
output is printed.
.PP
Format options:
.IP \[bu] 2
\f[B]table[:/path/to/file,\&...]\f[R]: Output events in table format.
The default path to the file is stdout.
Multiple file paths can be specified, separated by commas.
.IP \[bu] 2
\f[B]table-verbose[:/path/to/file,\&...]\f[R]: Output events in table
format with extra fields per event.
The default path to the file is stdout.
Multiple file paths can be specified, separated by commas.
.IP \[bu] 2
\f[B]json[:/path/to/file,\&...]\f[R]: Output events in JSON format.
The default path to the file is stdout.
Multiple file paths can be specified, separated by commas.
.IP \[bu] 2
\f[B]gob[:/path/to/file,\&...]\f[R]: Output events in gob format.
The default path to the file is stdout.
Multiple file paths can be specified, separated by commas.
.IP \[bu] 2
\f[B]gotemplate=/path/to/template[:/path/to/file,\&...]\f[R]: Output
events formatted using a given Go template file.
The default path to the file is stdout.
Multiple file paths can be specified, separated by commas.
.IP \[bu] 2
\f[B]none\f[R]: Ignore the stream of events output.
This is usually used with the \f[B]--capture\f[R] flag.
.PP
Fluent Forward options:
.IP \[bu] 2
\f[B]forward:url\f[R]: Send events in JSON format using the Forward
protocol to a Fluent receiver.
Specify the URL of the Fluent receiver.
.PP
Webhook options:
.IP \[bu] 2
\f[B]webhook:url\f[R]: Send events in JSON format to the specified
webhook URL.
.PP
Other options:
.IP \[bu] 2
\f[B]option:{stack-addresses,exec-env,relative-time,exec-hash,parse-arguments,sort-events}\f[R]:
Augment output according to the given options.
The default is none.
Multiple options can be specified, separated by commas.
.RS 2
.IP \[bu] 2
\f[B]stack-addresses\f[R]: Include stack memory addresses for each
event.
.IP \[bu] 2
\f[B]exec-env\f[R]: When tracing execve/execveat, show the environment
variables that were used for execution.
.IP \[bu] 2
\f[B]relative-time\f[R]: Use relative timestamp instead of wall
timestamp for events.
.IP \[bu] 2
\f[B]exec-hash\f[R]: When tracing sched_process_exec, show the file hash
(sha256) and ctime.
.IP \[bu] 2
\f[B]parse-arguments\f[R]: Do not show raw machine-readable values for
event arguments.
Instead, parse them into human-readable strings.
.IP \[bu] 2
\f[B]parse-arguments-fds\f[R]: Enable parse-arguments and enrich file
descriptors (fds) with their file path translation.
This can cause pipeline slowdowns.
.IP \[bu] 2
\f[B]sort-events\f[R]: Enable sorting events before passing them to the
output.
This may decrease the overall program efficiency.
.RE
.SS EXAMPLES
.IP \[bu] 2
To output events as JSON to stdout, use the following flag:
.RS 2
.IP
.nf
\f[C]
--output json
\f[R]
.fi
.RE
.IP \[bu] 2
To output events as JSON to \f[C]/my/out\f[R], use the following flag:
.RS 2
.IP
.nf
\f[C]
--output json:/my/out
\f[R]
.fi
.RE
.IP \[bu] 2
To output events as the provided Go template to stdout, use the
following flag:
.RS 2
.IP
.nf
\f[C]
--output gotemplate=/path/to/my.tmpl
\f[R]
.fi
.RE
.IP \[bu] 2
To output events in gob format to \f[C]/my/out\f[R], use the following
flag:
.RS 2
.IP
.nf
\f[C]
--output gob:/my/out
\f[R]
.fi
.RE
.IP \[bu] 2
To output events as JSON to stdout and as gob to \f[C]/my/out\f[R], use
the following flag:
.RS 2
.IP
.nf
\f[C]
--output json --output gob:/my/out
\f[R]
.fi
.RE
.IP \[bu] 2
To output events as JSON to both \f[C]/my/out\f[R] and
\f[C]/my/out2\f[R], use the following flag:
.RS 2
.IP
.nf
\f[C]
--output json:/my/out1,/my/out2
\f[R]
.fi
.RE
.IP \[bu] 2
To ignore events output, use the following flag:
.RS 2
.IP
.nf
\f[C]
--output none
\f[R]
.fi
.RE
.IP \[bu] 2
To output events as a table with stack addresses, use the following
flag:
.RS 2
.IP
.nf
\f[C]
--output table --output option:stack-addresses
\f[R]
.fi
.RE
.IP \[bu] 2
To output events via the Forward protocol to \f[C]127.0.0.1\f[R] on port
\f[C]24224\f[R] with the tag `tracker' using TCP, use the following
flag:
.RS 2
.IP
.nf
\f[C]
--output forward:tcp://user:pass\[at]127.0.0.1:24224?tag=tracker
\f[R]
.fi
.RE
.IP \[bu] 2
To output events to the webhook endpoint \f[C]http://webhook:8080\f[R],
use the following flag:
.RS 2
.IP
.nf
\f[C]
--output webhook:http://webhook:8080
\f[R]
.fi
.RE
.IP \[bu] 2
To output events to the webhook endpoint \f[C]http://webhook:8080\f[R]
with a timeout of 5 seconds, use the following flag:
.RS 2
.IP
.nf
\f[C]
--output webhook:http://webhook:8080?timeout=5s
\f[R]
.fi
.RE


pkg/cmd/flags/output.go
package flags

import (
	"errors"
	"fmt"
	"net/url"
	"os"
	"path/filepath"
	"strings"

	"github.com/khulnasoft-lab/tracker/pkg/config"
	"github.com/khulnasoft-lab/tracker/pkg/errfmt"
)

type PrepareOutputResult struct {
	TrackerConfig   *config.OutputConfig
	PrinterConfigs []config.PrinterConfig
}

func PrepareOutput(outputSlice []string, newBinary bool) (PrepareOutputResult, error) {
	outConfig := PrepareOutputResult{}
	trackerConfig := &config.OutputConfig{}

	// outpath:format
	printerMap := make(map[string]string)

	for _, o := range outputSlice {
		outputParts := strings.SplitN(o, ":", 2)

		if strings.HasPrefix(outputParts[0], "gotemplate=") {
			err := parseFormat(outputParts, printerMap, newBinary)
			if err != nil {
				return outConfig, err
			}
			continue
		}

		switch outputParts[0] {
		case "none":
			if len(outputParts) > 1 {
				if newBinary {
					// TODO: build man page
					// return outConfig, errors.New("none output does not support path. See 'tracker-output' man page for more info")
					return outConfig, errors.New("none output does not support path. Use '--help' for more info")
				}

				return outConfig, errors.New("none output does not support path. Use '--output help' for more info")
			}
			printerMap["stdout"] = "ignore"
		case "table", "table-verbose", "json", "gob":
			err := parseFormat(outputParts, printerMap, newBinary)
			if err != nil {
				return outConfig, err
			}
		case "forward":
			err := validateURL(outputParts, "forward", newBinary)
			if err != nil {
				return outConfig, err
			}

			printerMap[outputParts[1]] = "forward"
		case "webhook":
			err := validateURL(outputParts, "webhook", newBinary)
			if err != nil {
				return outConfig, err
			}

			printerMap[outputParts[1]] = "webhook"
		case "option":
			err := parseOption(outputParts, trackerConfig, newBinary)
			if err != nil {
				return outConfig, err
			}
		default:
			if newBinary {
				// TODO: build man page
				// return outConfig, fmt.Errorf("invalid output flag: %s, see 'tracker-output' man page for more info", outputParts[0])
				return outConfig, fmt.Errorf("invalid output flag: %s, use '--help' for more info", outputParts[0])
			}

			return outConfig, fmt.Errorf("invalid output flag: %s, use '--output help' for more info", outputParts[0])
		}
	}

	// default
	if len(printerMap) == 0 {
		printerMap["stdout"] = "table"
	}

	printerConfigs, err := getPrinterConfigs(printerMap, trackerConfig, newBinary)
	if err != nil {
		return outConfig, err
	}

	outConfig.TrackerConfig = trackerConfig
	outConfig.PrinterConfigs = printerConfigs

	return outConfig, nil
}

// setOption sets the given option in the given config
func setOption(cfg *config.OutputConfig, option string, newBinary bool) error {
	switch option {
	case "stack-addresses":
		cfg.StackAddresses = true
	case "exec-env":
		cfg.ExecEnv = true
	case "relative-time":
		cfg.RelativeTime = true
	case "exec-hash":
		cfg.ExecHash = true
	case "parse-arguments":
		cfg.ParseArguments = true
	case "parse-arguments-fds":
		cfg.ParseArgumentsFDs = true
		cfg.ParseArguments = true // no point in parsing file descriptor args only
	case "sort-events":
		cfg.EventsSorting = true
	default:
		if newBinary {
			// TODO: build man page
			// return errfmt.Errorf("invalid output option: %s, see 'tracker-output' man page for more info", option)
			return errfmt.Errorf("invalid output option: %s, use '--help' for more info", option)
		}

		return errfmt.Errorf("invalid output option: %s, use '--output help' for more info", option)
	}

	return nil
}

// getPrinterConfigs returns a slice of printer.Configs based on the given printerMap
func getPrinterConfigs(printerMap map[string]string, trackerConfig *config.OutputConfig, newBinary bool) ([]config.PrinterConfig, error) {
	printerConfigs := make([]config.PrinterConfig, 0, len(printerMap))

	for outPath, printerKind := range printerMap {
		if printerKind == "table" {
			if err := setOption(trackerConfig, "parse-arguments", newBinary); err != nil {
				return nil, err
			}
		}

		outFile := os.Stdout
		var err error

		if outPath != "stdout" && printerKind != "forward" && printerKind != "webhook" {
			outFile, err = createFile(outPath)
			if err != nil {
				return nil, err
			}
		}

		printerConfigs = append(printerConfigs, config.PrinterConfig{
			Kind:       printerKind,
			OutPath:    outPath,
			OutFile:    outFile,
			RelativeTS: trackerConfig.RelativeTime,
		})
	}

	return printerConfigs, nil
}

// parseFormat parses the given format and sets it in the given printerMap
func parseFormat(outputParts []string, printerMap map[string]string, newBinary bool) error {
	// if not file was passed, we use stdout
	if len(outputParts) == 1 {
		outputParts = append(outputParts, "stdout")
	}

	for _, outPath := range strings.Split(outputParts[1], ",") {
		if outPath == "" {
			if newBinary {
				// TODO: build man page
				// return errfmt.Errorf("format flag can't be empty, see 'tracker-output' man page for more info")
				return errfmt.Errorf("format flag can't be empty, use '--help' for more info")
			}

			return errfmt.Errorf("format flag can't be empty, use '--output help' for more info")
		}

		if _, ok := printerMap[outPath]; ok {
			if newBinary {
				// TODO: build man page
				// return errfmt.Errorf("cannot use the same path for multiple outputs: %s, see 'tracker-output' man page for more info", outPath)
				return errfmt.Errorf("cannot use the same path for multiple outputs: %s, use '--help' for more info", outPath)
			}

			return errfmt.Errorf("cannot use the same path for multiple outputs: %s, use '--output help' for more info", outPath)
		}
		printerMap[outPath] = outputParts[0]
	}

	return nil
}

// parseOption parses the given option and sets it in the given config
func parseOption(outputParts []string, trackerConfig *config.OutputConfig, newBinary bool) error {
	if len(outputParts) == 1 || outputParts[1] == "" {
		if newBinary {
			// TODO: build man page
			// return errfmt.Errorf("option flag can't be empty, see 'tracker-output' man page for more info")
			return errfmt.Errorf("option flag can't be empty, use '--help' for more info")
		}

		return errfmt.Errorf("option flag can't be empty, use '--output help' for more info")
	}

	for _, option := range strings.Split(outputParts[1], ",") {
		err := setOption(trackerConfig, option, newBinary)
		if err != nil {
			return err
		}
	}

	return nil
}

// creates *os.File for the given path
func createFile(path string) (*os.File, error) {
	fileInfo, err := os.Stat(path)
	if err == nil && fileInfo.IsDir() {
		return nil, errfmt.Errorf("cannot use a path of existing directory %s", path)
	}

	dir := filepath.Dir(path)
	if err := os.MkdirAll(dir, 0755); err != nil {
		return nil, errfmt.Errorf("failed to create directory: %v", err)
	}
	file, err := os.Create(path)
	if err != nil {
		return nil, errfmt.Errorf("failed to create output path: %v", err)
	}

	return file, nil
}

// validateURL validates the given URL
// --output [webhook|forward]:[protocol://user:pass@]host:port[?k=v#f]
func validateURL(outputParts []string, flag string, newBinary bool) error {
	if len(outputParts) == 1 || outputParts[1] == "" {
		if newBinary {
			// TODO: build man page
			// return errfmt.Errorf("%s flag can't be empty, see 'tracker-output' man page for more info", flag)
			return errfmt.Errorf("%s flag can't be empty, use '--help' for more info", flag)
		}

		return errfmt.Errorf("%s flag can't be empty, use '--output help' for more info", flag)
	}
	// Now parse our URL using the standard library and report any errors from basic parsing.
	_, err := url.ParseRequestURI(outputParts[1])

	if err != nil {
		if newBinary {
			// TODO: build man page
			// return errfmt.Errorf("invalid uri for %s output %q. Use 'tracker-output' man page for more info", flag, outputParts[1])
			return errfmt.Errorf("invalid uri for %s output %q. Use '--help' for more info", flag, outputParts[1])
		}

		return errfmt.Errorf("invalid uri for %s output %q. Use '--output help' for more info", flag, outputParts[1])
	}

	return nil
}
